validator_client/src/initialized_validators.rs

```diff
@@ -239,46 +241,45 @@ impl InitializedValidator {
                     voting_keypair: Arc::new(voting_keypair),
                 }
             }
-            SigningDefinition::Web3Signer {
-                url,
-                root_certificate_path,
-                request_timeout_ms,
-                client_identity_path,
-                client_identity_password,
-            } => {
-                let signing_url = build_web3_signer_url(&url, &def.voting_public_key)
+            SigningDefinition::Web3Signer(web3_signer) => {
+                let signing_url = build_web3_signer_url(&web3_signer.url, &def.voting_public_key)
                     .map_err(|e| Error::InvalidWeb3SignerUrl(e.to_string()))?;
-                let request_timeout = request_timeout_ms
+
+                let request_timeout = web3_signer
+                    .request_timeout_ms
                     .map(Duration::from_millis)
                     .unwrap_or(DEFAULT_REMOTE_SIGNER_REQUEST_TIMEOUT);
 
-                let builder = Client::builder().timeout(request_timeout);
-
-                let builder = if let Some(path) = root_certificate_path {
-                    let certificate = load_pem_certificate(path)?;
-                    builder.add_root_certificate(certificate)
+                // Check if a client has already been initialized for this remote signer url.
+                let http_client = if let Some(client_map) = web3_signer_client_map {
+                    match client_map.get(&web3_signer) {
+                        Some(client) => client.clone(),
+                        None => {
+                            let client = build_web3_signer_client(
+                                web3_signer.root_certificate_path.clone(),
+                                web3_signer.client_identity_path.clone(),
+                                web3_signer.client_identity_password.clone(),
+                                request_timeout,
+                            )?;
+                            client_map.insert(web3_signer, client.clone());
+                            client
+                        }
+                    }
```

Comment:
This pattern is captured well by the `HashMap::entry` API, which allows you to write:

```rust
client_map
    .entry(web3_signer)
    .or_insert_with(|| build_web3_signer_client(
        web3_signer.root_certificate_path.clone(),
        web3_signer.client_identity_path.clone(),
        web3_signer.client_identity_password.clone(),
        request_timeout,
    ))
    .clone()
```

It's a tiny bit shorter and slightly more efficient because it keeps a reference to the vacant hashmap entry after looking it up the first time

Reply:
I initially tried to use this pattern, but ran into issues since `build_web3_signer_client` returns a `Result` which couldn't be handled inside the closure.

Reply:
ah yeah! nevermind then!